src/commands/keygen.rs

```diff
@@ -0,0 +1,605 @@
+        long = "algorithm",
+        value_name = "ALGORITHM",
+        value_parser = ValueParser::new(Keygen::parse_algorithm),
+        verbatim_doc_comment,
+    )]
+    algorithm: GenerateParams,
+
+    /// Generate a key signing key instead of a zone signing key
+    #[arg(short = 'k')]
+    make_ksk: bool,
+
+    /// Whether to create symlinks.
+    #[arg(short, long, value_enum, num_args = 0..=1, require_equals = true, default_missing_value = "yes", default_value = "no")]
+    symlink: SymlinkArg,
+
+    /// The domain name to generate a key for
+    #[arg(value_name = "domain name", value_parser = ValueParser::new(parse_name))]
+    name: Name<Vec<u8>>,
+}
+
+/// Symlinking behaviour.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, ValueEnum)]
+pub enum SymlinkArg {
+    /// Don't create symlinks.
+    No,
+
+    /// Create symlinks, but don't overwrite existing ones.
+    Yes,
+
+    /// Create symlinks, overwriting existing ones.
+    Force,
+}
+
+impl SymlinkArg {
+    /// Whether symlinks should be created.
+    pub fn create(&self) -> bool {
+        matches!(self, Self::Yes | Self::Force)
+    }
+
+    /// Whether symlinks should be forced.
+    pub fn force(&self) -> bool {
+        matches!(self, Self::Force)
+    }
+}
+
+const LDNS_HELP: &str = "\
+ldns-keygen -a <algorithm> [-b bits] [-r /dev/random] [-s] [-f] [-v] domain
+  generate a new key pair for domain
+  -a <alg>	use the specified algorithm (-a list to show a list)
+  -k		set the flags to 257; key signing key
+  -b <bits>	specify the keylength (only used for RSA keys)
+  -r <random>	randomness device (unused)
+  -s		create additional symlinks with constant names
+  -f		force override of existing symlinks
+  -v		show the version and exit
+  The following files will be created:
+    K<name>+<alg>+<id>.key	Public key in RR format
+    K<name>+<alg>+<id>.private	Private key in key format
+    K<name>+<alg>+<id>.ds	DS in RR format (only for DNSSEC KSK keys)
+  The base name (K<name>+<alg>+<id>) will be printed to stdout
+";
+
+impl LdnsCommand for Keygen {
+    const HELP: &'static str = LDNS_HELP;
+
+    fn parse_ldns<I: IntoIterator<Item = OsString>>(args: I) -> Result<Self, Error> {
+        let mut algorithm = None;
+        let mut make_ksk = false;
+        let mut bits = 2048;
+        let mut create_symlinks = false;
+        let mut force_symlinks = false;
+        let mut name = None;
+
+        let mut parser = lexopt::Parser::from_args(args);
+
+        while let Some(arg) = parser.next()? {
+            match arg {
+                Arg::Short('a') => {
+                    if algorithm.is_some() {
+                        return Err("cannot specify algorithm (-a) more than once".into());
+                    }
+
+                    algorithm = parse_os_with("algorithm (-a)", &parser.value()?, |s| {
+                        Ok(match s {
+                            "list" => {
+                                // TODO: Mock stdout and process exit?
```

Comment:
This is possible after the notify branch is merged, where `parse_ldns` takes `Env` as well.